datumaro/components/format_detection.py

```diff
@@ -106,11 +116,18 @@ def _is_path_within_root(self, path: str) -> bool:
 
         return True
 
+    def _start_requirement(self, req_type: str) -> None:
+        assert not self._one_or_more_context, \
+            f"a requirement ({req_type}) can't be placed directly within " \
+            "a 'require_one_or_more' block"
+
     def fail(self, requirement: str) -> NoReturn:
         """
         Places a requirement that is never met. `requirement` must contain
         a human-readable description of the requirement.
         """
+        self._start_requirement("fail")
+
         raise FormatRequirementsUnmet((requirement,))
```

Comment:
```suggestion
    def fail(self, message: str) -> NoReturn:
        """
        Places a requirement that is never met. `requirement` must contain
        a human-readable description of the requirement.
        """
        self._start_requirement("fail")

        raise FormatRequirementsUnmet((message,))
```

Reply:
This sort of value is called `requirement_desc` elsewhere, so I renamed it to that instead.